common/src/types/v0/store/node.rs

```diff
@@ -1,95 +1,318 @@
+    pub fn add_drain_label(&mut self, label: &String) {
+        self.drainlabels.push(label.to_string());
+    }
+}
+
+/// Enum variant encompassing data related to a cordoned or draining node.
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
+pub enum CordonDrainState {
+    /// The node is being cordoned.
+    Cordoned(CordonedState),
+    /// The node is being drained.
+    Draining(DrainState),
+    /// The drain has completed.
+    Drained(DrainState),
+}
+
+/// Node information.
 #[derive(Serialize, Deserialize, Debug, PartialEq)]
 pub struct Node {
-    /// Node information.
+    /// Node state information.
     node: transport::NodeState,
     /// Node labels.
     labels: NodeLabels,
 }
 
+/// Node state information.
 pub struct NodeState {
-    /// Node information
+    /// Node information.
     pub node: transport::NodeState,
 }
 
+/// Node spec data, including the cordon/drain state.
 #[derive(Serialize, Deserialize, Debug, PartialEq, Default, Clone)]
 pub struct NodeSpec {
     /// Node identification.
     id: NodeId,
-    /// Endpoint of the io-engine instance (gRPC)
+    /// Endpoint of the io-engine instance (gRPC).
     endpoint: String,
     /// Node labels.
     labels: NodeLabels,
-    /// Cordon labels.
+    /// Cordon/drain state.
+    #[serde(skip_serializing_if = "Option::is_none")]
     #[serde(default)] // Ensure backwards compatibility in etcd when upgrading.
-    cordon_labels: Vec<String>,
+    cordon_drain_state: Option<CordonDrainState>,
 }
 
 impl NodeSpec {
-    /// Return a new `Self`
+    /// Return a new `Self`.
     pub fn new(
         id: NodeId,
         endpoint: String,
         labels: NodeLabels,
-        cordon_label: Option<Vec<String>>,
+        cordon_drain_state: Option<CordonDrainState>,
     ) -> Self {
         Self {
             id,
             endpoint,
             labels,
-            cordon_labels: cordon_label.unwrap_or_default(),
+            cordon_drain_state,
         }
     }
-    /// Node identification
+    /// Node identification.
     pub fn id(&self) -> &NodeId {
         &self.id
     }
-    /// Node gRPC endpoint
+    /// Node gRPC endpoint.
     pub fn endpoint(&self) -> &str {
         &self.endpoint
     }
-    /// Node labels
+    /// Node labels.
     pub fn labels(&self) -> &NodeLabels {
         &self.labels
     }
-    /// Node gRPC endpoint
+    /// Node labels.
+    pub fn cordon_drain_state(&self) -> &Option<CordonDrainState> {
+        &self.cordon_drain_state
+    }
+
+    /// Node gRPC endpoint.
     pub fn set_endpoint(&mut self, endpoint: String) {
         self.endpoint = endpoint
     }
+
     /// Cordon node by applying the label.
     pub fn cordon(&mut self, label: String) {
-        self.cordon_labels.push(label);
+        // the node has the label, return with an error
+        match &mut self.cordon_drain_state {
+            Some(ds) => match ds {
+                CordonDrainState::Cordoned(state) => {
```

Comment:
would it be tidier to add a add_cordon_label fn to CordonDrainState and handle this logic there?